compiler/src/dotty/tools/dotc/inlines/InlineReducer.scala

```diff
@@ -408,23 +395,39 @@ class InlineReducer(inliner: Inliner)(using Context):
             case ConstantValue(v: Boolean) => (v, true)
             case _ => (false, false)
           }
-        if guardOK then Some((caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to), canReduceGuard))
-        else if canReduceGuard then None
-        else Some((caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to), canReduceGuard))
+        if guardOK && canReduceGuard then MatchRedux.Reduced(caseBindings.map(_.subst(from, to)), cdef.body.subst(from, to))
+        else if canReduceGuard then MatchRedux.CaseNoMatch
+        else MatchRedux.NoCaseSelected
       }
-      else None
+      else MatchRedux.CaseNoMatch
     }
 
     def recur(cases: List[CaseDef]): MatchRedux = cases match {
-      case Nil => None
+      case Nil => MatchRedux.NoCaseSelected
       case cdef :: cases1 =>
         reduceCase(cdef) match
-          case None => recur(cases1)
-          case r @ Some((caseBindings, rhs, canReduceGuard)) if canReduceGuard => Some((caseBindings, rhs))
-          case _ => None
+          case MatchRedux.CaseNoMatch      => recur(cases1)
+          case reduced: MatchRedux.Reduced => reduced
+          case MatchRedux.NoCaseSelected   => MatchRedux.NoCaseSelected
     }
 
     recur(cases)
   }
 end InlineReducer
 
+/** The result type of reducing a match.
+ *  Either a list of bindings for the pattern-bound variables and the RHS of the selected case,
+ *  or NoCaseSelected. */
+enum MatchRedux extends MatchReduxWithCaseNoMatch:
+  case Reduced(caseBindings: List[MemberDef], rhs: Tree)
+  case NoCaseSelected
+object MatchRedux:
+  case object CaseNoMatch extends MatchReduxWithCaseNoMatch
```

Comment:
Why not make it an enum case? 

I find the logic with the two `MatchResult` types hard to understand.

Reply:
No functional changes intended at all.  `Reduced` and `NoCaseSelected` are the results for the reduction as a whole (`MatchRedux`), while `CaseNoMatch` is an added case for the recursion.  This allows the objects to be shared.  But I can change it to just leak out the `CaseNoMatch` to the callsite.  Or I can make two enums and rebox the result out.  Or keep this, but switch to a recognisable sealed/case hierarchy?  Or maybe better names would clear it up.  I'm not married to any particular encoding.